dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java

```diff
@@ -1455,75 +1454,129 @@ public boolean doesUserHavePermissions(PermissionableType permType, int permissi
-    	    				continue;
-    					}else{
-    						newSetOfPermissions.add(new Permission(permissionable.getPermissionId(), p.getRoleId(), p.getPermission(), true));
-    					}
-    				}
-    				if(!p.isIndividualPermission()){
-    				   newSetOfPermissions.add(new Permission(p.getType(), permissionable.getPermissionId(), p.getRoleId(), p.getPermission(), true));
-    				}
-    			}else{
-    				newSetOfPermissions.add(new Permission(p.getType(), permissionable.getPermissionId(), p.getRoleId(), p.getPermission(), true));
-    			}
-    		}
+        // We need to make sure that newSetOfPermissions doesn't contain
+        // a child of the role we are assigning permissions.
+        for (Permission newPermission : newSetOfPermissions) {
+            Role newPermissionRole = APILocator.getRoleAPI().loadRoleById(newPermission.getRoleId());
+
+            if (!APILocator.getRoleAPI().isParentRole(role, newPermissionRole)) {
+                newSetOfPermissionsFiltered.add(newPermission);
+            }
+        }
+
+        if (!newSetOfPermissions.isEmpty()) {
+            // NOTE: Method "assignPermissions" is deprecated in favor of "savePermission",
+            // which has subtle functional differences. Please take these differences into
+            // consideration if planning to replace this method with the "savePermission"
+            permissionFactory.assignPermissions(newSetOfPermissionsFiltered, permissionable);
+        }
+    }
 
+    /**
+     * Retrieves all the parent permissions in order to be applied to the permissionable.
+     */
+    private List<Permission> getNewPermissions(Permissionable parent, Permissionable permissionable,
+            User user) throws DotDataException, DotSecurityException {
+
+        List<Permission> perList = new ArrayList<>();
+        List<Permission> newSetOfPermissions = new ArrayList<>();
+
+        if (!doesUserHavePermission(permissionable, PermissionAPI.PERMISSION_EDIT_PERMISSIONS,
+                user)) {
+            throw new DotSecurityException("User id: " + user.getUserId()
+                    + " does not have permission to alter permissions on asset " + permissionable
+                    .getPermissionId());
+        }
+
+        if (parent.isParentPermissionable()) {
+
+            String type = permissionable.getPermissionType();
+            perList.addAll(permissionFactory.getInheritablePermissions(parent));
+            perList.addAll(permissionFactory.getPermissions(parent, true));
+            Host host = APILocator.getHostAPI()
+                    .find(permissionable.getPermissionId(), APILocator.getUserAPI().getSystemUser(), false);
+            if (host != null) {
+                type = Host.class.getCanonicalName();
+            }
+            for (Permission p : perList) {
+
+                if (type.equals(Folder.class.getCanonicalName())) {
+                    if (p.getType().equals(Template.class.getCanonicalName())

```

Comment:
You can encapsulate it on a set and do something like

private final Set<String> permissionType = set (Template.class.getCanonicalName(), Container.class.getCanonicalName() ....

Them, if if (type.equals(Folder.class.getCanonicalName()) && this.permissionType.contains(p.getType())) {